### ISSUE TITLE
A question about the efficiency of 2D sparse operations

### ISSUE THREAD

**Issue by lebron-2016** [ChenhongyiYang/QueryDet-PyTorch#98]
Dear author,

When I was testing the performance of the model, I was confused about the efficiency of the 2D sparse operation, so I did the following experiment and found that 2D sparse convolution does not improve the efficiency compared to normal convolution. Could you please explain what the problem is?

```
import spconv.pytorch as spconv
import torch
import torch.nn as nn
from spconv.pytorch import SparseConvTensor
import time

device='cuda:0'

x_d = torch.zeros((2, 4, 1024, 1024))
x_d[0,0,0:16,0:16] += 1.
x_d = x_d.to(device)
x = SparseConvTensor.from_dense(x_d.permute(0,2,3,1))

conv_sparse = spconv.SparseConv2d(4, 4, kernel_size=3,stride=2, padding=1,bias=False, dilation=1).to(device)
bn_sparse = nn.BatchNorm1d(4, momentum=0.1).to(device)
conv_bn_relu_sparse = spconv.SparseSequential(conv_sparse, bn_sparse, nn.ReLU(inplace=True)).to(device)

conv_norm = nn.Conv2d(4, 4, kernel_size=3,stride=2, padding=1,bias=False, dilation=1).to(device)
bn_norm = nn.BatchNorm2d(4, momentum=0.1).to(device)
conv_bn_relu_norm = nn.Sequential(conv_norm, bn_norm, nn.ReLU(inplace=True)).to(device)

start_time = time.time()
encoder_output1 = conv_bn_relu_norm(x_d)
part_time = time.time() - start_time
print("conv_bn_relu_norm time", part_time)
start_time = time.time()
encoder_output = conv_bn_relu_sparse(x)
part_time = time.time() - start_time
print("conv_bn_relu_sparse time", part_time)

start_time = time.time()
encoder_output1 = conv_bn_relu_norm(x_d)
part_time = time.time() - start_time
print("conv_bn_relu_norm time", part_time)
start_time = time.time()
encoder_output = conv_bn_relu_sparse(x)
part_time = time.time() - start_time
print("conv_bn_relu_sparse time", part_time)

start_time = time.time()
encoder_output1 = conv_bn_relu_norm(x_d)
part_time = time.time() - start_time
print("conv_bn_relu_norm time", part_time)
start_time = time.time()
encoder_output = conv_bn_relu_sparse(x)
part_time = time.time() - start_time
print("conv_bn_relu_sparse time", part_time)
```
![image](https://github.com/ChenhongyiYang/QueryDet-PyTorch/assets/80045218/223d63ff-174a-4fde-bc39-665ca5e3c973)

Looking forward to your reply!

Thanks a lot!!

**Comment by judycpChen** [comment]
hello,What is your spconv version

**Comment by lebron-2016** [comment]
> hello,What is your spconv version

Hi, my version is spconv-cu117(2.3.6). And I also try the version of spconv-cu102(2.1.25), the situation is similar.

I suspect there is something wrong with my usage here, could you please help me take a look?

Thanks a lot!!

**Comment by lebron-2016** [comment]
Hi, I am also a little confused. Why don't you use `spconv` directly during training, but perform weight transfer during testing?

Thanks!!

**Comment by ChenhongyiYang** [comment]
Hi, 

Thank you for the question. The reason is that using sparse conv during training will greatly reduce negative samples, which could cause training instability because we used focal loss for classification.

BW,
Chenhongyi
